File: 3227-find-missing-and-repeated-values/3227-find-missing-and-repeated-values.java
class Solution {
    public int[] findMissingAndRepeatedValues(int[][] arr) {
        int n = arr.length;
        int N = n * n;

        // Use long to prevent integer overflow
        long expectedSum = (long) N * (N + 1) / 2;
        long expectedSquareSum = (long) N * (N + 1) * (2 * N + 1) / 6;

        long actualSum = 0, actualSquareSum = 0;
        int[] count = new int[N + 1]; // Frequency array

        int repeated = -1, missing = -1;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                int num = arr[i][j];
                actualSum += num;
                actualSquareSum += (long) num * num;

                count[num]++;
                if (count[num] == 2) {
                    repeated = num;
                }
            }
        }

        long diff = expectedSum - actualSum; // missing - repeated
        missing = (int) (repeated + diff);

        return new int[]{repeated, missing};
    }
}

File: 3227-find-missing-and-repeated-values/3227-find-missing-and-repeated-values.cpp
class Solution {
public:
    vector<int> findMissingAndRepeatedValues(vector<vector<int>>& arr) {

        long long Sac = 0, Sex = 0, S2ac = 0, S2ex = 0;
        long long A, B;
        int n = arr.size() * arr.size();

        Sex = n * (n + 1) / 2ll;
        S2ex = n * (n + 1ll) * (2ll * n + 1ll) / 6ll;

        for (int i = 0; i < arr.size(); i++) {
            for (int j = 0; j < arr[i].size(); j++) {
                int t = arr[i][j];
                Sac = Sac + (long long)t;
                S2ac = S2ac + ((long long)t * (long long)t);
            }
        }

        A = Sex - Sac;
        B = (S2ex - S2ac) / A;

        int X = (A + B) / 2;
        int Y = (B - A) / 2;

        return {Y, X};
    }
};